### PR TITLE
Reconnect redux

### PR DESCRIPTION
**Pull request by bill-torpey**
# Connect/reconnect with 0mq

The purpose of this PR is to close the loop on connect/reconnect options that began with <https://github.com/zeromq/libzmq/pull/3831>.

We want to take advantage of 0mq's support for automatic reconnection of sockets that become disconnected, but there are some edge cases that are problematic for us.

Because of the way [OZ](https://github.com/nyfix/OZ/) does [dynamic discovery](https://github.com/nyfix/OZ/blob/master/doc/Naming-Service.md), processes don't bind to known ports -- (almost) all ports are assigned dynamically.  These port assignments are then published to the one-and-only known port, which then forwards them to all the other processes.

With automatic reconnection enabled, if a process terminates and restarts (e.g., because of a crash), 0mq will continue to try to reconnect to its original port.  

The scenarios we care about are:

1. Process terminates, peer attempts to reconnect but gets ECONNREFUSED.  This is already handled by the earlier PR.
2. Process mistakenly attempts to connect to a port that is not a 0mq process:
    1.  The non-0mq process unilaterally closes the connection, presumably because the 0mq process did not adhere to the protocol expected by the non-0mq process.
    2.  The non-0mq process does nothing -- i.e., it does not respond to the 0mq greeting.
    3.  The non-0mq process does respond in some fashion, or at least sends data on the connection that `receive_greeting` can read.

## 1. No process listening at port

This is handled by the ZMQ_RECONNECT_STOP/ZMQ_RECONNECT_STOP_CONN_REFUSED sockopt.

```
#0  zmq::socket_base_t::event_closed (this=0x614590, endpoint_uri_pair_=..., fd_=16) at /home/btorpey/work/libzmq/master/src/src/socket_base.cpp:1866
#1  0x00007ffff7b8f3cc in zmq::stream_connecter_base_t::close (this=0x7fffe80008c0) at /home/btorpey/work/libzmq/master/src/src/stream_connecter_base.cpp:158
#2  0x00007ffff7b68f90 in zmq::tcp_connecter_t::out_event (this=0x7fffe80008c0) at /home/btorpey/work/libzmq/master/src/src/tcp_connecter.cpp:110
#3  0x00007ffff7b8f438 in zmq::stream_connecter_base_t::in_event (this=0x7fffe80008c0) at /home/btorpey/work/libzmq/master/src/src/stream_connecter_base.cpp:168
#4  0x00007ffff7b266a0 in zmq::epoll_t::loop (this=0x60a170) at /home/btorpey/work/libzmq/master/src/src/epoll.cpp:198
#5  0x00007ffff7b42629 in zmq::worker_poller_base_t::worker_routine (arg_=0x60a170) at /home/btorpey/work/libzmq/master/src/src/poller_base.cpp:135
#6  0x00007ffff7b6a3ba in thread_routine (arg_=0x60a1c8) at /home/btorpey/work/libzmq/master/src/src/thread.cpp:250
#7  0x00007ffff7635e65 in start_thread () from /lib64/libpthread.so.0
#8  0x00007ffff6b3f88d in clone () from /lib64/libc.so.6
```

## 2. Non-0mq process listening at port
These all rely on the 0mq handshake failing.  There are a few ways that can happen.

### 2.i Peer closes connection
In some cases, the non-0mq process will unilaterally close the connection, presumably because the greeting from 0mq process violates the non-0mq process's expected behavior.  This can either be done gracefully (Oracle listener sends a `FIN`) or forcibly (smbd sends a `RST`) -- in either case it is detected in `receive_greeting` with errno == `ECONNRESET`.

### 2.ii Peer does nothing
Some processes don't respond at all to the 0mq greeting.  In this case, the handshake times out (in `stream_engine_base_t::timer_event`)
and triggers the disconnect:

```
#0  zmq::socket_base_t::event_handshake_failed_no_detail (this=0x614590, endpoint_uri_pair_=..., err_=11)
    at /home/btorpey/work/libzmq/master/src/src/socket_base.cpp:1887
#1  0x00007ffff7b66379 in zmq::stream_engine_base_t::error (this=0x7fffe8000f80, reason_=zmq::i_engine::timeout_error)
    at /home/btorpey/work/libzmq/master/src/src/stream_engine_base.cpp:689
#2  0x00007ffff7b66818 in zmq::stream_engine_base_t::timer_event (this=0x7fffe8000f80, id_=64)
    at /home/btorpey/work/libzmq/master/src/src/stream_engine_base.cpp:737
#3  0x00007ffff7b42473 in zmq::poller_base_t::execute_timers (this=0x60a170) at /home/btorpey/work/libzmq/master/src/src/poller_base.cpp:99
#4  0x00007ffff7b2653f in zmq::epoll_t::loop (this=0x60a170) at /home/btorpey/work/libzmq/master/src/src/epoll.cpp:173
#5  0x00007ffff7b42629 in zmq::worker_poller_base_t::worker_routine (arg_=0x60a170) at /home/btorpey/work/libzmq/master/src/src/poller_base.cpp:135
#6  0x00007ffff7b6a3be in thread_routine (arg_=0x60a1c8) at /home/btorpey/work/libzmq/master/src/src/thread.cpp:250
#7  0x00007ffff7635e65 in start_thread () from /lib64/libpthread.so.0
#8  0x00007ffff6b3f88d in clone () from /lib64/libc.so.6
```

This requires the fix described [below](#Possible-bug).

### 2.iii Peer responds to greeting

Some processes do respond to the 0mq greeting. (For instance, sshd responds to the greeting with "SSH-2.0-Open", which just happens to be the same length as the unversioned 0mq greeting).

#### Specifying ZMQ_MIN_ZMTP_VERSION
One way to detect/reject non-0mq processes is to specify a minimum ZMTP version (using the `ZMQ_MIN_ZMTP_VERSION` sockopt).  Since the greeting isn't a proper ZMTP greeting, the code treats it as "unversioned", and as such it doesn't match any ZMTP version.

```
#0  zmq::socket_base_t::event_disconnected (this=0x614590, endpoint_uri_pair_=..., fd_=16) at /home/btorpey/work/libzmq/master/src/src/socket_base.cpp:1880
#1  0x00007ffff7b66394 in zmq::stream_engine_base_t::error (this=0x7fffe8000f80, reason_=zmq::i_engine::protocol_error) at /home/btorpey/work/libzmq/master/src/src/stream_engine_base.cpp:694
#2  0x00007ffff7b8ccb5 in zmq::zmtp_engine_t::handshake_v1_0_unversioned (this=0x7fffe8000f80) at /home/btorpey/work/libzmq/master/src/src/zmtp_engine.cpp:260
#3  0x00007ffff7b8c668 in zmq::zmtp_engine_t::handshake (this=0x7fffe8000f80) at /home/btorpey/work/libzmq/master/src/src/zmtp_engine.cpp:133
#4  0x00007ffff7b64492 in zmq::stream_engine_base_t::in_event_internal (this=0x7fffe8000f80) at /home/btorpey/work/libzmq/master/src/src/stream_engine_base.cpp:252
#5  0x00007ffff7b643f2 in zmq::stream_engine_base_t::in_event (this=0x7fffe8000f80) at /home/btorpey/work/libzmq/master/src/src/stream_engine_base.cpp:242
#6  0x00007ffff7b26745 in zmq::epoll_t::loop (this=0x60a170) at /home/btorpey/work/libzmq/master/src/src/epoll.cpp:206
#7  0x00007ffff7b42629 in zmq::worker_poller_base_t::worker_routine (arg_=0x60a170) at /home/btorpey/work/libzmq/master/src/src/poller_base.cpp:135
#8  0x00007ffff7b6a3ba in thread_routine (arg_=0x60a1c8) at /home/btorpey/work/libzmq/master/src/src/thread.cpp:250
#9  0x00007ffff7635e65 in start_thread () from /lib64/libpthread.so.0
#10 0x00007ffff6b3f88d in clone () from /lib64/libc.so.6
```

#### Without ZMQ_MIN_ZMTP_VERSION
Last but not least is the case where a process responds to the 0mq greeting, but for whatever reason we don't want to use `ZMQ_MIN_ZMTP_VERSION`.  This is similar to the case 2.2 above, and requires the fix mentioned [below](#Possible-bug). 

## Possible bug
In this case, it looks like there's a bug in the code -- the following sequence should move from:

```
bool zmq::zmtp_engine_t::handshake ()
{
    ...
    if (_has_handshake_timer) {
        cancel_timer (handshake_timer_id);
        _has_handshake_timer = false;
    }

    return true;
}
```

to... 

```
void zmq::stream_engine_base_t::mechanism_ready ()
{
    ...
    if (_has_handshake_timer) {
        cancel_timer (handshake_timer_id);
        _has_handshake_timer = false;
    }

    _socket->event_handshake_succeeded (_endpoint_uri_pair, 0);
}
```

Testing appears to show that handshakes will never timeout with the current version of the code.

## Other Stuff

The doc for `ZMQ_EVENT_DISCONNECTED` states:

> The socket was disconnected unexpectedly. The event value is the FD of the underlying network socket. **Warning: this socket will be closed.**

That doesn't seem to happen, and I'm curious why?

## Test Code
Code used to run these tests is at: <https://github.com/WallStProg/zmqtests/tree/master/connect>.

### CODE REVIEW

File: src/zmtp_engine.cpp [hunk 1]
```diff
@@ -138,11 +137,6 @@ bool zmq::zmtp_engine_t::handshake ()
     if (_outsize == 0)
         set_pollout ();
 
-    if (_has_handshake_timer) {
```
Comment (somdoron) [inline src/zmtp_engine.cpp:141]: This is partially correct, for ZMTP version below 3 there is no mechanism_ready. You can follow the semantic of the following PR: https://github.com/zeromq/libzmq/pull/3913/files.

When engine_ready is called the timer should be canceled as well.

File: src/zmtp_engine.hpp [hunk 2]
```diff
@@ -119,9 +119,6 @@ class zmtp_engine_t ZMQ_FINAL : public stream_engine_base_t
     unsigned char _greeting_recv[v3_greeting_size];
     unsigned char _greeting_send[v3_greeting_size];
 
-    //  Size of greeting received so far
-    unsigned int _greeting_bytes_read;
```
Comment (somdoron) [inline src/zmtp_engine.hpp]: why was this moved to the base class?

**Comment by somdoron** [comment]
Looks good to me, the only comment is regarding canceling the timer on ZMTP versions below 3.

File: src/zmtp_engine.hpp (hunk 2 above)
Comment (bill-torpey) [reply src/zmtp_engine.hpp]: Corrected -- an earlier version of the code was checking for greeting_bytes_read > 0 in stream_engine_base.ccp, but that code was found to be superfluous.  See d9634e5b7bbc8168f568797fd0df0b94d1869913

File: src/zmtp_engine.cpp (hunk 1 above)
Comment (bill-torpey) [reply src/zmtp_engine.cpp:141]: Something I've been wondering is how to test with older versions of ZMTP?  Any suggestions for old versions of libzmq with which to test?  

I'm guessing that it would be pretty much impossible to test this as part of automated unit tests.

Comment (somdoron) [reply src/zmtp_engine.cpp:141]: You can test using the Stream socket or OS file descriptor, some examples here:
https://github.com/zeromq/libzmq/blob/master/tests/test_mock_pub_sub.cpp
https://github.com/zeromq/netmq/blob/master/src/NetMQ.Tests/ZMTPTests.cs#L38

Comment (bill-torpey) [reply src/zmtp_engine.cpp:141]: Right -- the updated test does exactly that to test connecting to non-0mq.  

The question is more about connecting to older versions of ZMTP -- afaict there is no way to force 0mq to present an older version of ZMTP to peers.
